### PR TITLE
feat(ai-ide): render delegated sessions as children in session list

### PR DESCRIPTION
**Pull request by ndoschek**
#### What it does

Group delegated chat sessions under their parent session in the welcome screen session list, as a collapsible multi-level hierarchy.

- Track each session's root and immediate parent on chat sessions and thread them through serialization, the chat service, and the session store so the welcome view can reconstruct the full delegation hierarchy.
- Render parent sessions with a collapsible expand/collapse toggle shown next to the agent icon; childless rows reserve an equal-sized placeholder so agent icons stay aligned across all rows.
- Indent child sessions per depth and give the expand toggle a larger hit area, a keyboard-only focus ring, selection-aware coloring, and an accessible label that names the session.
- Keep the Active and Restored sections in the welcome list, splitting top-level rows by section while children render nested under their parent.
- Auto-expand a session's ancestors when one of its delegated descendants starts awaiting user action, so it is revealed.
- Mark every ancestor of an awaiting session with the same bell icon, attention dot, and hover tooltip as the awaiting session — visible even on collapsed or not-yet-opened (restored) parents.
- Show delegated child sessions in the 'Browse all chats' quick input too (tagged as delegated), so every session stays reachable; the welcome list still nests them under their parent.
- Cascade session deletion to children, covering both in-memory and persisted-only children so a deleted parent never leaves orphans.

Note: I did a first few round with an AI review here: https://github.com/eclipsesource/theia/pull/323

#### How to test

1. Build (npm run build:browser) and start the browser app; create a delegated chat so a
  parent session gets child sessions.
2. In the welcome session list, confirm parents show the agent icon plus an expand chevron
  (childless rows keep icons aligned via a placeholder), and toggling shows/hides indented
  children.
3. Focus a parent row: the chevron stays visible on the selection background, and clicking it
  shows no focus ring.
4. Confirm both Active and Restored sections render; child sessions appear nested under their
  parent in the list, and also in the "Browse all chats" quick input tagged as *Delegated*.
5. Create a multi-level delegation (e.g. Coder → Architect → Explore) and confirm it nests
  recursively. When a nested session needs approval/input, confirm its ancestors auto-expand
  and show the bell icon plus attention dot, with a tooltip explaining a delegated session
  needs your attention.

#### Breaking changes

- [ ] This PR introduces breaking changes and requires careful review. If yes, the breaking changes section in the [changelog](https://github.com/eclipse-theia/theia/blob/master/CHANGELOG.md) has been updated.

#### Review checklist

- [x] As an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)
- [x] User-facing text is internationalized using the `nls` service (for details, please see the [Internationalization/Localization section](https://github.com/theia-ide/theia/blob/master/doc/coding-guidelines.md#internationalizationlocalization) in the Coding Guidelines)

#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

File: packages/ai-ide/src/browser/chat-session-item.tsx [hunk 1]
```diff
@@ -45,6 +45,19 @@ export interface ChatSessionItemProps {
     actions?: ChatSessionItemAction[];
     /** Invoked when an action is triggered, with the action and this item's session. */
     onAction?: (action: ChatSessionItemAction, session: ChatSessionMetadata) => void;
+    /**
+     * Whether this session has child sessions. The children themselves are rendered separately by
+     * the parent list; this only drives the expand/collapse toggle.
+     */
+    hasChildren?: boolean;
```
Comment (eneufeld) [inline packages/ai-ide/src/browser/chat-session-item.tsx]: i find the name `hasChildren` on a React Component not ideal. Maybe name it `hasChildSessions`?

File: packages/ai-ide/src/browser/chat-session-item.tsx [hunk 2]
```diff
@@ -45,6 +45,19 @@ export interface ChatSessionItemProps {
...
+    /** Whether this item is a child of a parent session. */
+    isChild?: boolean;
```
Comment (eneufeld) [inline packages/ai-ide/src/browser/chat-session-item.tsx]: Same as the `hasChildren` comment, maybe name `isChildSession`

File: packages/ai-ide/src/browser/chat-sessions-welcome-message-provider.tsx [hunk 3]
```diff
@@ -40,6 +40,13 @@ export interface SectionedSessions {
     restored: ChatSessionMetadata[];
 }
 
+/** A session row with its optional children. */
+export interface SessionRow {
+    session: ChatSessionMetadata;
+    isRestored: boolean;
+    children: SessionRow[];
```
Comment (eneufeld) [inline packages/ai-ide/src/browser/chat-sessions-welcome-message-provider.tsx]: Same as above, imho it should be childSessions

File: packages/ai-ide/src/browser/chat-sessions-welcome-message-provider.tsx [hunk 4]
```diff
@@ -356,41 +379,96 @@ export class ChatSessionsWelcomeMessageProvider implements ChatWelcomeMessagePro
...
+        // Build one row per session (carrying its own restored flag), then nest each child row under
+        // its root session's row via rootSessionId. Children keep their row so the restored flag and
+        // actions are computed once here rather than recomputed per child during rendering.
+        const allSessions = [...sections.active, ...sections.restored];
+        const activeIds = new Set(sections.active.map(s => s.sessionId));
+        const rowsById = new Map<string, SessionRow>();
+        for (const session of allSessions) {
+            rowsById.set(session.sessionId, { session, isRestored: !activeIds.has(session.sessionId), children: [] });
+        }
+        for (const session of allSessions) {
```
Comment (eneufeld) [inline packages/ai-ide/src/browser/chat-sessions-welcome-message-provider.tsx:417]: why two loops?

**Review by eneufeld — COMMENTED** [review]
Code looks fine, need to test yet. Only some naming comments

File: packages/ai-ide/src/browser/chat-session-item.tsx (hunk 1 above)
Comment (ndoschek) [reply packages/ai-ide/src/browser/chat-session-item.tsx]: Great catch, yes definitely, I updated to `hasChildSessions` and `isChildSession`

File: packages/ai-ide/src/browser/chat-sessions-welcome-message-provider.tsx (hunk 4 above)
Comment (ndoschek) [reply packages/ai-ide/src/browser/chat-sessions-welcome-message-provider.tsx:417]: I added a comment there to explain why

File: packages/ai-chat/src/common/chat-service.ts [hunk 5]
```diff
@@ -224,6 +224,31 @@ export class ChatServiceImpl implements ChatService {
...
+            }
+        }
+        if (this.sessionStore) {
+            try {
+                const index = await this.sessionStore.getSessionIndex();
+                for (const metadata of Object.values(index)) {
+                    if (metadata.rootSessionId === sessionId) {
+                        childIds.add(metadata.sessionId);
+                    }
+                }
```
Comment (eneufeld) [inline packages/ai-chat/src/common/chat-service.ts:257]: Is it intentional that root deletion continues when the index read fails? Persisted-only children cannot be discovered in that case, so deleting the root can leave orphaned child sessions.

File: packages/ai-chat-ui/src/browser/ai-chat-ui-contribution.ts [hunk 6]
```diff
@@ -448,7 +448,7 @@ export class AIChatContribution extends AbstractViewContribution<ChatViewWidget>
     protected async askForChatSession(): Promise<QuickPickItem | undefined> {
         const getItems = async (): Promise<(QuickPickItem | QuickPickSeparator)[]> => {
             const activeSessions = this.chatService.getSessions()
-                .filter(session => session.title)
+                .filter(session => session.title && !session.rootSessionId)
```
Comment (eneufeld) [inline packages/ai-chat-ui/src/browser/ai-chat-ui-contribution.ts]: Was it intentional that child sessions become unreachable when `WELCOME_SCREEN_SESSIONS_PREF` is `0`? The inline hierarchy is hidden in that configuration, and this filter also removes children from the fallback picker.

File: packages/ai-ide/src/browser/chat-session-item.tsx [hunk 7]
```diff
@@ -213,26 +231,52 @@ export function ChatSessionItem(props: ChatSessionItemComponentProps): React.Rea
...
+                        onKeyDown={e => {
+                            if (isActivationKey(e)) {
+                                e.preventDefault();
+                                e.stopPropagation();
+                                onToggleExpand?.();
+                            }
+                        }}
+                        onMouseDown={e => e.preventDefault()}
+                        role="button"
+                        tabIndex={0}
```
Comment (eneufeld) [inline packages/ai-ide/src/browser/chat-session-item.tsx:264]: Every toggle is announced only as "Expand" or "Collapse", so a screen-reader user cannot tell which session it controls. Include the session title in the accessible label.

**Review by eneufeld — APPROVED** [review]
Looks good to me, come clarifying questions left

**Comment by eneufeld** [comment]
Hmm I ran a test. I like the drop down, I down't like that notifications are not propagated anymore (at least I think they were).
Screenshot from 2026-07-15 13-50-17
Screenshot from 2026-07-15 13-50-20

File: packages/ai-chat-ui/src/browser/ai-chat-ui-contribution.ts (hunk 6 above)
Comment (ndoschek) [reply packages/ai-chat-ui/src/browser/ai-chat-ui-contribution.ts]: Ah indeed, good catch thanks!
I though again about this and rather than filtering child sessions out of the picker (which made them unreachable when the welcome list is hidden), the picker now shows every session and tags delegated children with a "Delegated" marker in the description. Reachability no longer depends on the welcome-screen session limit.

File: packages/ai-ide/src/browser/chat-session-item.tsx (hunk 7 above)
Comment (ndoschek) [reply packages/ai-ide/src/browser/chat-session-item.tsx:264]: The toggle's aria-label now includes the session title, e.g. "Expand {title}" / "Collapse {title}".

File: packages/ai-chat/src/common/chat-service.ts (hunk 5 above)
Comment (ndoschek) [reply packages/ai-chat/src/common/chat-service.ts:257]: Yes, this is intentional: the cascade delete is best-effort (the root's own storage delete already swallows errors too). Orphaned children aren't stranded: the home view would promote them to top-level and the picker now lists every session.

**Comment by ndoschek** [comment]
> Hmm I ran a test. I like the drop down, I down't like that notifications are not propagated anymore (at least I think they were).

Oh indeed that should definitely be propagated, I'll look into it!

**Comment by ndoschek** [comment]
I added another commit on top that auto expands parent sessions if one of the child sessions need attention in some way.

File: packages/ai-ide/src/browser/chat-sessions-welcome-message-provider.tsx [hunk 8]
```diff
@@ -286,13 +286,24 @@ export class ChatSessionsWelcomeMessageProvider implements ChatWelcomeMessagePro
...
             unread: false,
+            requiresAction: ChatSessionStatus.requiresUserAction(session.model.status),
             seenRequests: reqs.length,
             seenCompleted: this.countCompleted(reqs),
             listener: new DisposableCollection()
         };
         this.unreadSessions.set(session.id, state);
 
         session.model.onDidChange(() => {
+            const requiresAction = ChatSessionStatus.requiresUserAction(session.model.status);
```
Comment (eneufeld) [inline packages/ai-ide/src/browser/chat-sessions-welcome-message-provider.tsx:306]: not sure this will always work correctly, what about a multi hierarchy delegation?
eg Coder->Architect->Explore 
So either rekursive logic here or imho if possible move the notification to the top level similar to how we do it in the delegateAgentRenderer in the chat

Comment (ndoschek) [reply packages/ai-ide/src/browser/chat-sessions-welcome-message-provider.tsx:306]: Thanks, good catch! It i did not support deeper hierarchies at all, but this use case you mention is valid, I just haven't tested it yet.

So I added now deeper hierarchies and parent sessions that have any child sessions that need attention also show the bell icon and a proper attention label in the tooltip:

image

File: packages/ai-ide/src/browser/chat-sessions-welcome-message-provider.tsx [hunk 9]
```diff
@@ -264,13 +295,24 @@ export class ChatSessionsWelcomeMessageProvider implements ChatWelcomeMessagePro
...
+            if (requiresAction !== state.requiresAction) {
```
Comment (eneufeld) [inline packages/ai-ide/src/browser/chat-sessions-welcome-message-provider.tsx:309]: the check is not very obvious. but fine for me. ideally there would be a check on the state or session if a child needs attention or maybe add a short comment what this checks

**Review by eneufeld — APPROVED** [review]
Looks good to me. The remaining comment is a small nitpick that can be ignored.

File: packages/ai-ide/src/browser/chat-sessions-welcome-message-provider.tsx [hunk 10]
```diff
@@ -356,41 +400,117 @@ export class ChatSessionsWelcomeMessageProvider implements ChatWelcomeMessagePro
...
+     * Expands every ancestor of the given session (walking the immediate-parent chain) so a deeply
+     * delegated session becomes visible. The caller is responsible for triggering a re-render.
+     */
+    protected expandAncestors(session: ChatSession): void {
+        const visited = new Set<string>();
+        let parentId = session.parentSessionId ?? session.rootSessionId;
+        while (parentId && !visited.has(parentId)) {
+            visited.add(parentId);
+            this.expandedRoots.add(parentId);
+            const parent = this.chatService.getSession(parentId);
```
Comment (eneufeld) [inline packages/ai-ide/src/browser/chat-sessions-welcome-message-provider.tsx]: This walk stops as soon as an ancestor is persisted-only because `getSession` only returns loaded sessions. If a child is opened directly from Browse all chats, a later attention state expands its immediate parent but not the root, so the child remains hidden. Use the persisted metadata as a fallback while walking ancestors.

File: packages/ai-chat/src/common/chat-service.ts [hunk 11]
```diff
@@ -224,6 +226,31 @@ export class ChatServiceImpl implements ChatService {
     }
 
     async deleteSession(sessionId: string): Promise<void> {
+        // Delete children (sessions whose rootSessionId points to this session) first. Children may
+        // live only in memory, only in persisted storage (e.g. after a reload, not yet restored),
+        // or both, so collect ids from both sources to avoid orphaning persisted-only children.
+        const childIds = new Set<string>();
+        for (const s of this._sessions) {
+            if (s.rootSessionId === sessionId) {
```
Comment (eneufeld) [inline packages/ai-chat/src/common/chat-service.ts]: This only finds sessions whose `rootSessionId` equals the deleted ID. Deleting an intermediate session in A → B → C leaves C behind because C's root is A and its parent is B. Include immediate children via `parentSessionId`, then recurse with a visited set.

File: packages/ai-ide/src/browser/chat-session-item.tsx [hunk 12]
```diff
@@ -213,26 +239,55 @@ export function ChatSessionItem(props: ChatSessionItemComponentProps): React.Rea
...
+                        }}
+                        onKeyDown={e => {
+                            if (isActivationKey(e)) {
+                                e.preventDefault();
+                                e.stopPropagation();
+                                onToggleExpand?.();
+                            }
+                        }}
+                        onMouseDown={e => e.preventDefault()}
+                        role="button"
```
Comment (eneufeld) [inline packages/ai-ide/src/browser/chat-session-item.tsx:263]: This focusable button role is nested inside the focusable row button. Nested interactive roles are ambiguous for assistive technology. Give the hierarchy tree/treeitem semantics or make the row and toggle sibling controls.

File: packages/ai-ide/src/browser/chat-sessions-welcome-message-provider.tsx [hunk 13]
```diff
@@ -356,41 +400,117 @@ export class ChatSessionsWelcomeMessageProvider implements ChatWelcomeMessagePro
...
+        const allSessions = [...sections.active, ...sections.restored];
+        const activeIds = new Set(sections.active.map(s => s.sessionId));
+        // Two passes on purpose: the first pass creates a row for every session so the second pass can
+        // link each child to its parent. Sessions are ordered by recency, not parent-first, so a child
+        // may precede its parent here; linking in a single pass would miss parents not yet created.
+        const rowsById = new Map<string, SessionRow>();
+        for (const session of allSessions) {
+            rowsById.set(session.sessionId, { session, isRestored: !activeIds.has(session.sessionId), childSessions: [] });
+        }
+        for (const session of allSessions) {
```
Comment (eneufeld) [inline packages/ai-ide/src/browser/chat-sessions-welcome-message-provider.tsx:417]: This adds multi-level tree construction and recursive attention propagation without focused tests. Add coverage for a restored A → B hierarchy with active C, including ancestor expansion when C requires action.

**Review by eneufeld — COMMENTED** [review]
Some corner cases that I missed before

**Comment by ndoschek** [comment]
Thanks for the review @eneufeld!
I pushed another commit addressing the review comments: fixed the cascade delete for nested sessions (A → B → C), made expandAncestors fall back to the persisted index so directly-opened children still reveal their chain, and added tests for the multi-level cases.

The nested interactive-roles comment snowballed into proper keyboard navigation for the session list, which is well out of scope here, so I'll do that as a dedicated follow-up PR for keyboard navigation & accessibility improvements.

**Review by eneufeld — APPROVED** [review]
LGTM! Thank you
